pkg/virtctl/expose/expose.go

```diff
@@ -59,262 +61,236 @@ Possible types are (case insensitive, both single and plurant forms):
-	case "ExternalName":
-		return fmt.Errorf("type: %s not supported", strServiceType)
-	default:
-		return fmt.Errorf("unknown service type: %s", strServiceType)
+	var err error
+	if c.namespace, _, err = c.clientConfig.Namespace(); err != nil {
+		return err
+	}
+	if c.client, err = kubecli.GetKubevirtClientFromClientConfig(c.clientConfig); err != nil {
+		return fmt.Errorf("cannot obtain KubeVirt client: %v", err)
 	}
 
-	ipFamilies, err := convertIPFamily(strIPFamily)
+	serviceSelector, ports, err := c.getServiceSelectorAndPorts(vmType, vmName)
 	if err != nil {
 		return err
 	}
 
-	ipFamilyPolicy, err := convertIPFamilyPolicy(strIPFamilyPolicy, ipFamilies)
-	if err != nil {
+	if err := c.createService(serviceSelector, ports); err != nil {
 		return err
 	}
 
-	// get the namespace
-	namespace, _, err := o.clientConfig.Namespace()
-	if err != nil {
+	cmd.Printf("Service %s successfully created for %s %s\n", c.serviceName, vmType, vmName)
+	return nil
+}
+
+func (c *command) parseFlags() error {
+	c.targetPort = intstr.Parse(c.strTargetPort)
+
+	var err error
+	if c.protocol, err = convertProtocol(c.strProtocol); err != nil {
 		return err
 	}
-
-	// get the client
-	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
-	if err != nil {
-		return fmt.Errorf("cannot obtain KubeVirt client: %v", err)
+	if c.serviceType, err = convertServiceType(c.strServiceType); err != nil {
+		return err
+	}
+	if c.ipFamilies, err = convertIPFamily(c.strIPFamily); err != nil {
+		return err
+	}
+	if c.ipFamilyPolicy, err = convertIPFamilyPolicy(c.strIPFamilyPolicy, c.ipFamilies); err != nil {
+		return err
 	}
 
-	// does a plain quorum read from the apiserver
-	options := k8smetav1.GetOptions{}
+	return nil
+}
+
+func (c *command) getServiceSelectorAndPorts(vmType, vmName string) (map[string]string, []k8sv1.ServicePort, error) {
 	var serviceSelector map[string]string
-	ports := []v1.ServicePort{}
+	var ports []k8sv1.ServicePort
 
 	switch vmType {
 	case "vmi", "vmis", "virtualmachineinstance", "virtualmachineinstances":
-		// get the VM
-		vmi, err := virtClient.VirtualMachineInstance(namespace).Get(context.Background(), vmName, options)
+		vmi, err := c.client.VirtualMachineInstance(c.namespace).Get(context.Background(), vmName, metav1.GetOptions{})
 		if err != nil {
-			return fmt.Errorf("error fetching VirtualMachineInstance: %v", err)
+			return nil, nil, fmt.Errorf("error fetching VirtualMachineInstance: %v", err)
 		}
-		serviceSelector = vmi.ObjectMeta.Labels
 		ports = podNetworkPorts(&vmi.Spec)
-		// remove unwanted labels
-		delete(serviceSelector, virtv1.NodeNameLabel)
-		delete(serviceSelector, virtv1.VirtualMachinePoolRevisionName)
-		delete(serviceSelector, virtv1.MigrationTargetNodeNameLabel)
+		serviceSelector = vmi.ObjectMeta.Labels
+		delete(serviceSelector, v1.NodeNameLabel)
+		delete(serviceSelector, v1.VirtualMachinePoolRevisionName)
+		delete(serviceSelector, v1.MigrationTargetNodeNameLabel)
```

Comment:
nit - Just pull the labels we need from the VMI?

Reply:
IIUC the idea behind this is that we don't necessarily know which labels the VM/VMI/VMIRS will have. But we know which labels we don't want to have in the label selector.

Reply:
The VM might have no labels at all.

Reply:
Sounds like a sensible change, but I think it is out of scope for this refactor.

@EdDev Do you see any issues with this approach? I'd create a follow up PR for it.

Reply:
Lets continue this on a follow up PR please :slightly_smiling_face:

Reply:
> @EdDev Do you see any issues with this approach? I'd create a follow up PR for it.

I'm a bit more aggressive on this.
I am not interested to support `virtctl` shortcuts like this `expose` command.
Therefore, if we cannot remove it, then lets just lower our investment in it to the minimum.

Said that, if you think that the investment is worth it, please go ahead and push it forward.
I have not enough knowledge to know if the suggestion raised here is safe. But if we change something like this, we seem to need the e2e back :/.

Reply:
To be honest, it looks like it was working before and nobody complained. Right now the tool does not make assumptions about labels on a Pod and instead it errors out if there are no labels left to build a selector from. In the interest of keeping the investment to the minimum I would suggest to keep the current behavior.